### PR TITLE
`<iterator>`: Modernize and deprecate `(un)checked_array_iterator`

### LINKED ISSUE #943: <iterator>: checked_array_iterator/unchecked_array_iterator should have conversions
**Describe the bug**
The non-Standard extensions `stdext::checked_array_iterator` and `stdext::unchecked_array_iterator` don't participate in constness conversions like modern iterators do:

https://github.com/microsoft/STL/blob/5ef22f2a884706442257cea3dd9d5a6186f15c07/stl/inc/iterator#L521-L523
https://github.com/microsoft/STL/blob/5ef22f2a884706442257cea3dd9d5a6186f15c07/stl/inc/iterator#L718-L720

**Command-line test case**
```
C:\Temp>type meow.cpp
#include <array>
#include <iostream>
#include <iterator>
#include <type_traits>
using namespace std;
using stdext::checked_array_iterator;
using stdext::unchecked_array_iterator;

#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)

template <typename From, typename To>
void verify_convertible() {
    STATIC_ASSERT(is_convertible_v<From, To>);
    STATIC_ASSERT(is_constructible_v<To, From>);
    STATIC_ASSERT(is_assignable_v<To&, From>);
}

template <typename From, typename To>
void verify_NOT_convertible() {
    STATIC_ASSERT(!is_convertible_v<From, To>);
    STATIC_ASSERT(!is_constructible_v<To, From>);
    STATIC_ASSERT(!is_assignable_v<To&, From>);
}

struct Base {};
struct Derived : Base {};

int main() {
    array<int, 5> arr{{10, 20, 30, 40, 50}};

    {
        array<int, 5>::iterator standard_iter = arr.begin();
        standard_iter += 1;
        cout << *standard_iter << "\n";

        array<int, 5>::const_iterator standard_const_iter = arr.cbegin();
        standard_const_iter += 2;
        cout << *standard_const_iter << "\n";

        standard_const_iter = standard_iter;
        cout << *standard_const_iter << "\n";

        verify_convertible<array<int, 5>::iterator, array<int, 5>::const_iterator>();
        verify_NOT_convertible<array<Derived, 5>::iterator, array<Base, 5>::iterator>();
    }

    {
        checked_array_iterator<int*> checked_iter{arr.data(), arr.size()};
        checked_iter += 1;
        cout << *checked_iter << "\n";

        checked_array_iterator<const int*> checked_const_iter{arr.data(), arr.size()};
        checked_const_iter += 2;
        cout << *checked_const_iter << "\n";

#ifndef AVOID_ERRORS
        checked_const_iter = checked_iter;
        cout << *checked_const_iter << "\n";

        verify_convertible<checked_array_iterator<int*>, checked_array_iterator<const int*>>();
        verify_NOT_convertible<checked_array_iterator<Derived*>, checked_array_iterator<Base*>>();
#endif // AVOID_ERRORS
    }

    {
        unchecked_array_iterator<int*> unchecked_iter{arr.data()};
        unchecked_iter += 3;
        cout << *unchecked_iter << "\n";

        unchecked_array_iterator<const int*> unchecked_const_iter{arr.data()};
        unchecked_const_iter += 4;
        cout << *unchecked_const_iter << "\n";

#ifndef AVOID_ERRORS
        unchecked_const_iter = unchecked_iter;
        cout << *unchecked_const_iter << "\n";

        verify_convertible<unchecked_array_iterator<int*>, unchecked_array_iterator<const int*>>();
        verify_NOT_convertible<unchecked_array_iterator<Derived*>, unchecked_array_iterator<Base*>>();
#endif // AVOID_ERRORS
    }
}

C:\Temp>cl /EHsc /nologo /W4 /DAVOID_ERRORS meow.cpp && meow
meow.cpp
20
30
20
20
30
40
50

C:\Temp>cl /EHsc /nologo /W4 meow.cpp && meow
meow.cpp
meow.cpp(57): error C2679: binary '=': no operator found which takes a right-hand operand of type 'stdext::checked_array_iterator<int *>' (or there is no acceptable conversion)
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\iterator(710): note: could be 'stdext::checked_array_iterator<const int *> &stdext::checked_array_iterator<const int *>::operator =(stdext::checked_array_iterator<const int *> &&)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\iterator(710): note: or       'stdext::checked_array_iterator<const int *> &stdext::checked_array_iterator<const int *>::operator =(const stdext::checked_array_iterator<const int *> &)'
meow.cpp(57): note: while trying to match the argument list '(stdext::checked_array_iterator<const int *>, stdext::checked_array_iterator<int *>)'
meow.cpp(75): error C2679: binary '=': no operator found which takes a right-hand operand of type 'stdext::unchecked_array_iterator<int *>' (or there is no acceptable conversion)
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\iterator(850): note: could be 'stdext::unchecked_array_iterator<const int *> &stdext::unchecked_array_iterator<const int *>::operator =(stdext::unchecked_array_iterator<const int *> &&)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\iterator(850): note: or       'stdext::unchecked_array_iterator<const int *> &stdext::unchecked_array_iterator<const int *>::operator =(const stdext::unchecked_array_iterator<const int *> &)'
meow.cpp(75): note: while trying to match the argument list '(stdext::unchecked_array_iterator<const int *>, stdext::unchecked_array_iterator<int *>)'
```

**Expected behavior**
Code should compile.

**STL version**
Visual Studio Community 2019 Preview
16.7.0 Preview 3.1

**Additional context**
This item is also tracked on Developer Community as DevCom-241515 and by Microsoft-internal VSO-146139 / [AB#146139](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/146139).

### LINKED ISSUE #951: <iterator>: Consider deprecating and removing stdext::checked_array_iterator/unchecked_array_iterator
We've been shipping the non-Standard extensions `stdext::checked_array_iterator` and `stdext::unchecked_array_iterator` for a long time (I believe since VS 2005):

https://github.com/microsoft/STL/blob/5ef22f2a884706442257cea3dd9d5a6186f15c07/stl/inc/iterator#L521-L523
https://github.com/microsoft/STL/blob/5ef22f2a884706442257cea3dd9d5a6186f15c07/stl/inc/iterator#L718-L720

As non-Standard extensions, they're rarely used. They haven't been updated to behave like modern iterators (see #943) and they supported a design that we've ripped out (where the STL algorithms used to emit warnings when raw pointers were used as output iterators; this has been superseded by static analysis). Finally, they have been essentially superseded by better types: `std::span` (which provides debug checks and release performance) and `gsl::span` (which provides unconditional security checks).

I believe that we should consider marking these `stdext` iterators as deprecated in VS 2019, and remove them entirely in `vNext`.

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #943. Fixes #951.

This PR
- introduces convertibility from `(un)checked_array_iterator<T*>` to `(un)checked_array_iterator<const T*>`,
- deprecates `(un)checked_array_iterator` and `make_(un)checked_array_iterator` in C++17 and later modes, while allowing users to define `_SILENCE_STDEXT_ARR_ITERS_DEPRECATION_WARNING` to suppress the deprecation,
- diagnoses `(un)checked_array_iterator<void*>` and `(un)checked_array_iterator<Ret(*)(Args...)>` in `static_assert`,
- expands `iterator_traits<Ptr>::meow` typedefs,
- uses default member initializers for default construction,
- strengthens exception specification for `make_checked_array_iterator`, and
- ensures that `(un)checked_array_iterator` are contiguous iterators in C++20.

### CODE REVIEW

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/iterator [hunk 1]
```diff
@@ -1471,23 +1471,36 @@ using _STD iterator_traits;
...
-    using difference_type   = typename iterator_traits<_Ptr>::difference_type;
-    using pointer           = typename iterator_traits<_Ptr>::pointer;
-    using reference         = typename iterator_traits<_Ptr>::reference;
+    using iterator_category = _STD random_access_iterator_tag;
+    using value_type        = _STD remove_cv_t<_Pointee_type>;
+    using difference_type   = _STD ptrdiff_t;
+    using pointer           = _Ptr;
+    using reference         = _Pointee_type&;
+#ifdef __cpp_lib_concepts
+    using iterator_concept = _STD contiguous_iterator_tag;
```
[the remark this answers is not in the document]
Comment (frederick-vs-ja) [reply stl/inc/iterator:1487]: I'm now choosing making `checked_array_iterator` model `contiguous_iterator`...

**Review by CaseyCarter — APPROVED** [review]

**Review by strega-nil-ms — APPROVED** [review]

File: stl/inc/iterator [hunk 2]
```diff
@@ -1802,15 +1830,35 @@ public:
...
     return unchecked_array_iterator<_Ptr>(_It);
 }
+_STL_RESTORE_DEPRECATED_WARNING
 _STDEXT_END
 
+#if _HAS_CXX20
+_STD_BEGIN
+_STL_DISABLE_DEPRECATED_WARNING
+template <class _Ty>
+struct _CXX_DEPRECATE_STDEXT_ARR_ITERS pointer_traits<_STDEXT checked_array_iterator<_Ty*>> {
```
Comment (StephanTLavavej) [inline stl/inc/iterator:1849]: Deprecating this partial specialization doesn't seem right. The class template and factory function are deprecated, so any attempt to introduce usage will be diagnosed. This partial specialization of `pointer_traits` will typically be used by library code, which is just attempting to handle a `checked_array_iterator` given to it from the user-programmer.

File: tests/std/tests/Dev10_709166_checked_and_unchecked_array_iterator/test.cpp [hunk 3]
```diff
@@ -1,15 +1,71 @@
...
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
+
+template <class T>
+void check_checked_array_iterator_category_and_convertibility() {
+    STATIC_ASSERT(std::is_same_v<typename stdext::checked_array_iterator<T*>::iterator_category,
+        std::random_access_iterator_tag>);
+
+    STATIC_ASSERT(std::is_same_v<typename stdext::checked_array_iterator<T*>::value_type, std::remove_cv_t<T>>);
+
+    STATIC_ASSERT(std::is_same_v<typename stdext::checked_array_iterator<T*>::difference_type, std::ptrdiff_t>);
```
Comment (StephanTLavavej) [inline tests/std/tests/Dev10_709166_checked_and_unchecked_array_iterator/test.cpp:22]: Should include `<cstddef>` for `std::ptrdiff_t`.

File: tests/std/tests/Dev10_709166_checked_and_unchecked_array_iterator/test.cpp [hunk 4]
```diff
@@ -1,15 +1,71 @@
...
+
+    STATIC_ASSERT(std::contiguous_iterator<stdext::unchecked_array_iterator<T*>>);
+#endif // __cpp_lib_concepts
+}
+
 int main() {
     {
+        check_checked_array_iterator_category_and_convertibility<int>();
+        check_checked_array_iterator_category_and_convertibility<const int>();
+        check_checked_array_iterator_category_and_convertibility<std::tuple<const char*>>();
```
Comment (StephanTLavavej) [inline tests/std/tests/Dev10_709166_checked_and_unchecked_array_iterator/test.cpp:66]: Need to include `<tuple>` for `std::tuple`.

File: tests/std/tests/Dev10_709166_checked_and_unchecked_array_iterator/test.cpp [hunk 5]
```diff
@@ -19,31 +75,28 @@ int main() {
...
-        static_assert(std::is_same_v<decltype(dog), stdext::checked_array_iterator<int*>>,
-            "stdext::make_checked_array_iterator(p, 9, 3)'s return type is wrong!");
-
+        STATIC_ASSERT(std::is_same_v<decltype(cat), stdext::checked_array_iterator<int*>>);
 
-        static_assert(
-            std::is_same_v<stdext::checked_array_iterator<int*>::iterator_category, std::random_access_iterator_tag>,
-            "stdext::checked_array_iterator<int *>::iterator_category is wrong!");
+#if _HAS_CXX20
+        assert(std::to_address(cat) == &*cat);
```
Comment (StephanTLavavej) [inline tests/std/tests/Dev10_709166_checked_and_unchecked_array_iterator/test.cpp:81]: Need to include `<memory>` for `std::to_address`.

**Review by StephanTLavavej — COMMENTED** [review]
Thanks! I'll validate and push changes.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
FYI @strega-nil-ms I pushed changes after you approved; probably the most significant being the removal of deprecation on the `pointer_traits` partial specialization.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for enhancing and then discouraging use of this old machinery! :derelict_house: :house_with_garden: :sparkles:

**Pull request merged**